src/snowflake/snowpark/mock/plan.py

```diff
@@ -895,6 +909,215 @@ def calculate_expression(
+                        f"[Local Testing] Mocking window function {window_function.name} is not implemented."
+                    )
+                to_pass_args = []
+                for idx, key in enumerate(signatures.parameters):
+                    if key == spec.varargs:
+                        to_pass_args.extend(evaluated_children[idx:])
+                    else:
+                        try:
+                            to_pass_args.append(evaluated_children[idx])
+                        except IndexError:
+                            to_pass_args.append(None)
+                res_cols.append(
+                    _MOCK_FUNCTION_IMPLEMENTATION_MAP[window_function.name](
+                        *to_pass_args
+                    )
+                )
+            res_col = pd.concat(res_cols)
+            res_col.index = res_index
+            return res_col.sort_index()
+        elif isinstance(window_function, (Lead, Lag)):
+            offset = window_function.offset * (
+                1 if isinstance(window_function, Lead) else -1
+            )
+            ignore_nulls = window_function.ignore_nulls
+            res_cols = []
+            for current_row, w in zip(res_index, windows):
+                row_idx = list(w.index).index(
+                    current_row
+                )  # the row's 0-base index in the window
+                offset_idx = row_idx + offset
+                if offset_idx < 0 or offset_idx >= len(w):
+                    res_cols.append(
+                        calculate_expression(
+                            window_function.default,
+                            w,
+                            analyzer,
+                            expr_to_alias,
+                            keep_literal=True,
+                        )
+                    )
+                elif not ignore_nulls or offset == 0:
+                    res_cols.append(
+                        calculate_expression(
+                            window_function.expr,
+                            w.iloc[offset_idx],
+                            analyzer,
+                            expr_to_alias,
+                            keep_literal=True,
+                        )
+                    )
+                else:
+                    # skip rows where expr is NULL
+                    delta = 1 if offset > 0 else -1
+                    cur_idx = row_idx + delta
+                    cur_count = 0
+                    while 0 <= cur_idx < len(w):
+                        target_expr = calculate_expression(
+                            window_function.expr,
+                            w.iloc[cur_idx],
+                            analyzer,
+                            expr_to_alias,
+                            keep_literal=True,
+                        )
+                        if target_expr is not None:
+                            cur_count += 1
+                            if cur_count == abs(offset):
```

Comment:
I have a dumb question: why we stop here when `cur_count == abs(offset)`, this offset is the start offset of the window or the end offset of the window?

Reply:
Negative offset just means moving in the other direction. That's why I define `delta` and take the absolute value